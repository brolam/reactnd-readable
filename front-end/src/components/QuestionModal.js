import React from 'react'
import PropTypes from 'prop-types'

let questionModalTimeout

function QuestionModal(props) {
  let isYesOrNoClicked = false
  const { message, onYesAnswer, onNoAnswer, timeout } = props
  doNoAnswerEventWhenTimeout(timeout, onNoAnswer)
  return (
    <div className="modal-short modal-open"
      onClick={e => {
        if (isYesOrNoClicked) return
        clearTimeoutAndDoEvent(e, onNoAnswer)
      }
      }>
      <div className="modal-short-dialog">
        <h1>
          {message}
          <a
            className="yes"
            onClick={e => {
              isYesOrNoClicked = true
              clearTimeoutAndDoEvent(e, onYesAnswer)
            }
            }>Yes? </a>
          or <a
            className="no"
            onClick={e => {
              isYesOrNoClicked = true
              clearTimeoutAndDoEvent(e, onNoAnswer)
            }}>No?</a>
        </h1>
      </div>
    </div>
  )
}

function clearTimeoutAndDoEvent(e, eventTag) {
  clearTimeout(questionModalTimeout)
  eventTag(e)
}

function doNoAnswerEventWhenTimeout(timeout, onNoAnswer) {
  questionModalTimeout = setTimeout(() => {
    onNoAnswer();
  }, timeout);
}

QuestionModal.propTypes = {
  message: PropTypes.string.isRequired,
  onYesAnswer: PropTypes.func.isRequired,
  onNoAnswer: PropTypes.func.isRequired,
  timeout: PropTypes.number.isRequired
}

export default QuestionModal