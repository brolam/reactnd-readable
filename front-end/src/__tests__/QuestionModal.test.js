import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import QuestionModal from '../components/QuestionModal'

test('render without error', () => {
  ReactDOM.render(
    <QuestionModal
      message={"A question was render without error"}
      onYesAnswer={e => { }}
      onNoAnswer={e => { }}
      timeout={3000}
    />, document.createElement('div'));
})

test('last Snapshot', () => {
  const questionModal = renderer.create(
    <QuestionModal
      message={"Question?"}
      onYesAnswer={e => { }}
      onNoAnswer={e => { }}
      timeout={3000}
    />);
  let tree = questionModal.toJSON();
  expect(tree).toMatchSnapshot();
})

test('when clicking anywhere and not on yes or no', () => {
  let isClikedNoAnswer = false;
  const questionModal = mount(
    <QuestionModal
      message={"Question?"}
      onYesAnswer={e => { }}
      onNoAnswer={e => { isClikedNoAnswer = true }}
      timeout={3000}
    />);
  questionModal.find('div [className="modal-short"]').simulate('click')
  expect(isClikedNoAnswer).toBe(true);
})

test('close automatic on default timeout', () => {
  jest.useFakeTimers();
  let isClikedNoAnswer = false;
  const questionModal = mount(
    <QuestionModal
      message={"Question?"}
      onYesAnswer={e => { }}
      onNoAnswer={e => { isClikedNoAnswer = true }}
      timeout={3000}
    />);
  setTimeout(() => {
    expect(isClikedNoAnswer).toBe(true);
  }, 5000);
  jest.runAllTimers();
})

test('on click yes answer', () => {
  let isClikedYesAnswer = false;
  const questionModal = mount(
    <QuestionModal
      message={"Question?"}
      onYesAnswer={e => { isClikedYesAnswer = true }}
      onNoAnswer={e => { }}
      timeout={3000}
    />);
  questionModal.find('a [className="yes"]').simulate('click')
  expect(isClikedYesAnswer).toBe(true);
})

test('on click no answer', () => {
  let isClikedNoAnswer = false;
  const questionModal = mount(
    <QuestionModal
      message={"Question?"}
      onYesAnswer={e => { }}
      onNoAnswer={e => { isClikedNoAnswer = true }}
      timeout={3000}
    />);
  questionModal.find('a [className="no"]').simulate('click')
  expect(isClikedNoAnswer).toBe(true);
})