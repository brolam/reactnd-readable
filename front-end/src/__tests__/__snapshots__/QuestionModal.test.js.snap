// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`last Snapshot 1`] = `
<div
  className="modal-short modal-open"
  onClick={[Function]}
>
  <div
    className="modal-short-dialog"
  >
    <h1>
      Question?
      <a
        className="yes"
        onClick={[Function]}
      >
        Yes? 
      </a>
      or 
      <a
        className="no"
        onClick={[Function]}
      >
        No?
      </a>
    </h1>
  </div>
</div>
`;
